README.md
# Month Picker Strip
Widget is a slider that helps to pick month.

![](example/dev/screenshots/month_strip.gif)

# Usage
Let's demo the basic usage
```dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import 'package:month_picker_strip/month_picker_strip.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    const title = 'Month Strip Demo';
    return new MaterialApp(
      title: title,
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: title),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final DateFormat dateFormat = new DateFormat('MMMM yyyy');
  DateTime selectedMonth;

  @override
  void initState() {
    super.initState();
    selectedMonth = new DateTime(2017, 6);
  }

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        title: new Text(widget.title),
      ),
      body: new SafeArea(
        child: new Column(
          children: <Widget>[
            new Divider(
              height: 1.0,
            ),
            new MonthStrip(
              format: 'MMM yyyy',
              from: new DateTime(2016, 4),
              to: new DateTime(2018, 5),
              initialMonth: selectedMonth,
              height: 48.0,
              viewportFraction: 0.25,
              onMonthChanged: (v) {
                setState(() {
                  selectedMonth = v;
                });
              },
            ),
            new Divider(
              height: 1.0,
            ),
            new Expanded(
              child: new Center(
                child: new Text(dateFormat.format(selectedMonth)),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```


CHANGELOG.md
## 0.0.4-nullsafety.0

* Null Safety Migration

## 0.0.3

* Downgraded intl version to work it with current flutter_localizations package

## 0.0.2

* Added metadata for required fields


## 0.0.1

Initial Version of the library.
* Includes basic level widget


example/test/widget_test.dart
// This is a basic Flutter widget test.
// To perform an interaction with a widget in your test, use the WidgetTester utility that Flutter
// provides. For example, you can send tap and scroll gestures. You can also use WidgetTester to
// find child widgets in the widget tree, read text, and verify that the values of widget properties
// are correct.

import 'package:example/main.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(new MyApp());


  });
}


lib/month_picker_strip.dart
library month_picker_strip;

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:intl/intl.dart';

const PageScrollPhysics _kPagePhysics = const PageScrollPhysics();

const TextStyle _selectedTextStyle = const TextStyle(
  color: const Color(0xFF35567D),
  fontSize: 14.0,
  fontWeight: FontWeight.w600,
);

const TextStyle _normalTextStyle = const TextStyle(
  color: const Color(0x7F000000),
  fontSize: 14.0,
  fontWeight: FontWeight.w400,
);

class MonthStrip extends StatefulWidget {
  final String format;

  final DateTime from;
  final DateTime to;
  final DateTime initialMonth;
  final ValueChanged<DateTime> onMonthChanged;
  final double height;
  final double viewportFraction;
  final TextStyle selectedTextStyle;
  final TextStyle normalTextStyle;

  /// Defaults to matching platform conventions.
  final ScrollPhysics? physics;

  const MonthStrip({
    this.format = 'MMMM yyyy',
    required this.from,
    required this.to,
    required this.initialMonth,
    required this.onMonthChanged,
    this.physics,
    this.height = 48.0,
    this.viewportFraction = 0.3,
    this.normalTextStyle = _normalTextStyle,
    this.selectedTextStyle = _selectedTextStyle,
  });

  @override
  _MonthStripState createState() {
    List<_MonthItem> months = [];
    int initialPage = 0;
    for (int i = from.year; i <= to.year; i++) {
      for (int j = 1; j <= 12; j++) {
        if (i == from.year && j < from.month) {
          continue;
        }

        if (i == to.year && j > to.month) {
          continue;
        }

        var item = new _MonthItem(new DateTime(i, j), selected: false);
        if (item.time.year == initialMonth.year && item.time.month == initialMonth.month) {
          initialPage = months.length;
          item.selected = true;
        }
        months.add(item);
      }
    }

    return new _MonthStripState(
      viewportFraction: viewportFraction,
      initialPage: initialPage,
      dateFormat: new DateFormat(format),
      months: months,
    );
  }
}

class _MonthStripState extends State<MonthStrip> {
  final DateFormat dateFormat;
  final List<_MonthItem> months;
  final PageController controller;
  int _lastReportedPage;

  _MonthStripState({
    required double viewportFraction,
    required int initialPage,
    required this.dateFormat,
    required this.months,
  })   : controller = new PageController(
          viewportFraction: viewportFraction,
          initialPage: initialPage,
        ),
        _lastReportedPage = initialPage;

  @override
  Widget build(BuildContext context) {
    final AxisDirection axisDirection = AxisDirection.right;
    final ScrollPhysics physics = _kPagePhysics.applyTo(widget.physics);
    return new Container(
      height: widget.height,
      child: new NotificationListener<ScrollEndNotification>(
        onNotification: (ScrollEndNotification notification) {
          if (notification.depth == 0 && notification is ScrollEndNotification) {
            final PageMetrics metrics = notification.metrics as PageMetrics;
            final int currentPage = metrics.page!.round();
            if (currentPage != _lastReportedPage) {
              _lastReportedPage = currentPage;

              setState(() {
                for (var item in months) {
                  item.selected = false;
                }
                var m = months[currentPage];
                var d = m.time;
                m.selected = true;
                widget.onMonthChanged(new DateTime(d.year, d.month));
              });
            }
          }
          return false;
        },
        child: new Scrollable(
          axisDirection: axisDirection,
          controller: controller,
          physics: physics,
          viewportBuilder: (BuildContext context, ViewportOffset position) {
            return new Viewport(
              axisDirection: axisDirection,
              offset: position,
              slivers: <Widget>[
                new SliverFillViewport(
                  viewportFraction: controller.viewportFraction,
                  delegate:
                      new SliverChildBuilderDelegate(_buildContent, childCount: months.length),
                ),
              ],
            );
          },
        ),
      ),
    );
  }

  Widget _buildContent(BuildContext context, int index) {
    final item = months[index];
    return new Container(
      child: new Center(
        child: new GestureDetector(
          child: new Text(
            dateFormat.format(item.time),
            style: item.selected ? widget.selectedTextStyle : widget.normalTextStyle,
          ),
          onTap: () {
            if (_lastReportedPage != index) {
              controller.animateToPage(
                index,
                duration: new Duration(milliseconds: 300),
                curve: Curves.ease,
              );
            }
          },
        ),
      ),
    );
  }
}

class _MonthItem {
  final DateTime time;
  bool selected;

  _MonthItem(this.time, {this.selected = false});
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import 'package:month_picker_strip/month_picker_strip.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    const title = 'Month Strip Demo';
    return new MaterialApp(
      title: title,
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: title),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final DateFormat dateFormat = new DateFormat('MMMM yyyy');
  late DateTime selectedMonth;

  @override
  void initState() {
    super.initState();
    selectedMonth = new DateTime(2017, 6);
  }

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        title: new Text(widget.title),
      ),
      body: new SafeArea(
        child: new Column(
          children: <Widget>[
            new Divider(
              height: 1.0,
            ),
            new MonthStrip(
              format: 'MMM yyyy',
              from: new DateTime(2016, 4),
              to: new DateTime(2018, 5),
              initialMonth: selectedMonth,
              height: 48.0,
              viewportFraction: 0.25,
              onMonthChanged: (v) {
                setState(() {
                  selectedMonth = v;
                });
              },
            ),
            new Divider(
              height: 1.0,
            ),
            new Expanded(
              child: new Center(
                child: new Text(dateFormat.format(selectedMonth)),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
